APPS-1121 Update integration test suite

Gvaihir opened the pull request:
(feat) all - initial

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) all - initial

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) all - mostly unit test set up

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) all - TDD scaffold

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) all - WIP

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) TestDiscovery.py - add_tests method

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) TestDiscovery.py - done, tested

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) TestDiscovery.py - added language def for detection
(feat) utils.py - utils
(feat) RegisteredTest.py - added main props

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Context.py - common context of the execution
(feat) RegisteredTest.py - WIP

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Terraform.py - WIP

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Context.py - immutable + tests
(feat) Terraform.py - WIP + tests

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Context.py - extended functionality
(feat) Terraform.py - WIP
(feat) Dependency.py - Initial

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Context.py - extended functionality
(feat) Terraform.py - done, untested.
(feat) Dependency.py - WIP

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Dependency.py - done, untested
(fix) Terraform.py - methods moved to Dependency class
(feat) TestDiscovery.py - discovery of dependencies

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Terraform.py - done, untested

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Dependency.py - cast as a factory to account for different dependency types
(feat) TestDiscovery.py - Incorporated DependencyFactory
(feat) Terraform.py - common method to export dependency to dxasset.json

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) Context.py - fixed directories + tests
(feat) Dependency.py - comments

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Dependency.py - comments

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Messenger.py - untested
(feat) RegisteredTest.py - WIP

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - WIP
(fix) TestDiscovery.py - added Context
(fix) conftest.py - moved fixtures

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) TestDiscovery.py - fixture for Dependency tests

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Dependency.py - tested, done

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Terraform.py - WIP tests

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Context.py - refactoring and cleanup func
(feat) Terraform.py - finished, 95% tested

Gvaihir commented:
At the moment if reviewing pls focus on `Terraform` and `Dependency` classes and their tests. This is just to address [APPS-1122](https://jira.internal.dnanexus.com/browse/APP) ticket

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) test_Terraform.py - removed redundant tests.

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - implemented base class

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - factory object

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - finished base class. Not tested
(fix) Terraform.py - moved async retry to utils.py

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - wdl stub

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - tested bas class

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) TestDiscovery.py - handling single test name

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) cli.py - finished

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) cli.py - added arg
(fix) Context.py - repo root is an arg

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py & Terraform.py - added logging

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) cli.py - logger format
(feat) LogFormatter.py - class for logger formatting
(feat) RegisteredTest.py - validator method

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) cli.py - add tests CLI

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) README.md - documentation
(feat) cli.py - suites
(feat) Context.py - empty context

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) dependency configs

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) dependency configs

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Dependency.py - tarball dependency class + test. Removed update dot env.

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) pilot.json - pilot config for new tests

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) cli.py - docstring

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) Context.py - minor refactor

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) TestDiscovery.py - test category collection handling

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) added tests to the suite

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) Logger.py - common logger for the app
(feat) Context.py - common logger is in the context

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) everywhere - propagating common logger

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) everywhere - propagating common logger

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) Logger.py - fixed log duplication

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) minor logging fixes

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) ExpectedFailure.py - implementation of failing tests
(feat) RegisteredTestFactory.py - refactored, moved to a separate module

Gvaihir commented:
Ready for review. Also implemented expected failure tests. As an e2e demo feel free to run 

```

dxcint integration ${COMPILER_REPO_ROOT} -t "bad_status"
  
  ```

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - handling test inputs + tests

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) RegisteredTest.py - handling test inputs + tests

YuxinShi0423 commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+## Installing
+1. Clone this repo
+2. `cd dxcint`
+3. `potry install` (recommended, because will be installed from `poetry.lock` with pinned dependency versions). Or 
```
```suggestion
3. `poetry install` (recommended, because will be installed from `poetry.lock` with pinned dependency versions). Or 
```

YuxinShi0423 commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+
+## CLI methods
+#### Adding tests
+See [Adding tests](#adding-tests) section
```
the link here is pointing to itself...same as the running tests below

Gvaihir replied:
fixed in [f935157](https://github.com/dnanexus/dxCompiler/pull/312/commits/f93515739a8f2a8e04b66487c02cbdddbe0b9bfd)
Good catch, thanks!

YuxinShi0423 commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+
+Options:
+  -d, --directory TEXT  Directory containing workflow fixtures to be added
+  -x, --extension TEXT  Extension of the test workflow script files. Allowed
```
Would all these options be optional? 
Also maybe the extension could be parsed from the filename instead?

Gvaihir replied:
Nope, but I provided some default values in docstring. Moved some to positional args (category and suite names).
Extension is to explicitly control which formats (languages) you are adding for testing. It also is used for extraction of test names.

YuxinShi0423 commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+the names in appropriate config file.
+
+## Extending the framework
+### Adding new test types
```
```suggestion
### Adding new test categories
```

Gvaihir replied:
Categories yes, but here types is a more general definition that includes new implementation (subclass) and a new category name. Example here: https://github.com/dnanexus/dxCompiler/blob/aabdbce620bb6e0d60a19a95d3fe255627d58d7c/dxcint/dxcint/RegisteredTestFactory.py#L17
there might be a new category, but implementation may be one of the existing subclasses of `RegisteredTest`. Therefore I wanted to separate those definitions

YuxinShi0423 reviewed (commented):
partially reviewed. Will give dxCint a try! Cool work!

YuxinShi0423 commented on dxcint/dependencies/config/node.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "name": "node",
```
nodejs?

Gvaihir replied:
right, that's for `nodejs` but when you unpack the binary, it's located in `tarball/bin/node` - this is to avoid too much of hardcoding and just use a dependency name. Would be essential to consider how that naming works if we add new dependencies as tarball dependency.

Gvaihir committed:
Update dxcint/README.md

Co-authored-by: Yuxin Shi <yshi@dnanexus.com>

Gvaihir committed:
Update dxcint/README.md

Co-authored-by: Yuxin Shi <yshi@dnanexus.com>

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) README.md - links

Gvaihir committed:
Merge remote-tracking branch 'origin/APPS-1121_integration_test_suite' into APPS-1121_integration_test_suite

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) README.md - links

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) cli.py - move some options to positional args

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) RegisteredTestFactory.py - minor reformatting

YuxinShi0423 commented on dxcint/dxcint/Context.py:
```diff
@@ -0,0 +1,153 @@
+    def _resolve_project(project: str) -> str:
+        # First, see if the project is a project-id.
+        try:
+            project = dxpy.DXProject(project)
+            return project.get_id()
```
I think it is better not to reuse the`project` var name. maybe use `found_project` instead?

Gvaihir replied:
thanks, renamed in [cb81b09](https://github.com/dnanexus/dxCompiler/pull/312/commits/cb81b094cc5073b769b829ca017d221a92fb4b99)

YuxinShi0423 commented on dxcint/dxcint/Dependency.py:
```diff
@@ -0,0 +1,200 @@
+
+        src_link = config.get("source_link", None)
+        if src_link:
+            if re.search(r"tar\..z$", src_link):
```
so only the ones with extension `tar.gz` but not any others?

Gvaihir replied:
Actually `tar.xz` as well, but added regex to recognize `bz2` in [cb81b09](https://github.com/dnanexus/dxCompiler/pull/312/commits/cb81b094cc5073b769b829ca017d221a92fb4b99)

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) Context.py - minor renaming.
(fix) Dependency.py - regex to recognize bz2

commandlinegirl commented on dxcint/README.md:
````diff
@@ -0,0 +1,112 @@
+{"Suite name": "config file name in the dxcint/config where test names are registered"}
+```
+```commandline
+Usage: dxcint suites [OPTIONS]
````
Is there something similar for categories? No worries if not. If not perhaps adding a note where they are listed or a few examples (so that a new developer could grep them and orient themselves in the tests and understand what a category is) to line 26 would do.

Gvaihir replied:
I actually moved `category` to the positional args, will update the doc

Gvaihir replied:
Added docstring in [47da8f7](https://github.com/dnanexus/dxCompiler/pull/312/commits/47da8f742f5ec0f13dcaf060d19ec1c30e5adc2e).
Adding this functionality will require a user to provide a repo root like in other CLI functions. Suite names are hardcoded, so it's unnecessary for getting them.

commandlinegirl commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+Then, if this category needs a new behavior when compiling, running and/or validating the test, a new subclass of 
+`RegisteredTest` needs to be implemented. There might be up to 3 methods that might need implementation.
+
+1. **Compilation**. If a test workflows in this new test category should be compiled with a specific combination of dxCompiler flags, 
```
typo: s/workflows/workflow

commandlinegirl commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+The following architecture will be implemented to support integration tests for the compiler.
+1. **CLI** orchestrates the flow 
+2. **TestDiscovery** scans for config files, registers the tests by creating an array of RegisteredTest objects. 
+3. **Dependency** class sets up necessary environment with immutable state for **RegisteredTest**’s.
```
Nit: no apostrophe at the end

commandlinegirl commented on dxcint/README.md:
```diff
@@ -0,0 +1,112 @@
+**!!!Please consult the IT and compliance teams before adding new dependencies**
+#### Structure
+**Required fields**:
+1. "name": str. Name of a dependency. E.g. “awscli”
```
Nit: please update quotes here and in other places from ” to \"

Gvaihir replied:
fix in [5b1d4f4](https://github.com/dnanexus/dxCompiler/pull/312/commits/5b1d4f4f93bf0e44ade813517f2a57e2e9c58421)

commandlinegirl commented on dxcint/dependencies/config/dxfuse.json:
```diff
@@ -0,0 +1,6 @@
+  "name": "dxfuse",
+  "languages": ["wdl", "cwl"],
+  "version": "v1.0.0",
+  "source_link": "https://github.com/dnanexus/dxfuse/releases/download/v1.0.0/dxfuse-linux"
```
The most recent version is now 1.0.1

Gvaihir replied:
fixed in [5b1d4f4](https://github.com/dnanexus/dxCompiler/pull/312/commits/5b1d4f4f93bf0e44ade813517f2a57e2e9c58421)

commandlinegirl replied:
Sorry, in the meantime we had another release 1.1.0 :)

Gvaihir replied:
fixed in [ccbd011](https://github.com/dnanexus/dxCompiler/pull/312/commits/ccbd011d62b8468a944ee7b99644f65c8797e7dc)

commandlinegirl commented on dxcint/dependencies/config/cwltool.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "cwltool",
+  "languages": ["cwl"],
+  "version": "3.1.20220623174452",
```
Will these scripts be used in the release? https://github.com/dnanexus/dxCompiler/blob/85f5deaa590125bcceea08786e598bbb5be9f15d/.github/workflows/release.yml#L68?

Gvaihir replied:
Possibly. I would gradually migrate to `dxcint` for all such things - wherever possible for the sake of unified processes and documentation.

commandlinegirl commented on dxcint/dxcint/Context.py:
```diff
@@ -0,0 +1,153 @@
+        """
+        subdirectories = (os.path.join(base_dir, x) for x in ("applets", "test"))
+        for subdir in subdirectories:
+            _ = project_new_folder(
```
Curious what the purpose of the assignment is here

Gvaihir replied:
Explicit throwaway variable. `project_new_folder` eventually returns an API response.

commandlinegirl commented on dxcint/dxcint/Dependency.py:
```diff
@@ -0,0 +1,200 @@
+
+        src_link = config.get("source_link", None)
+        if src_link:
+            if re.search(r"tar\..z2?$", src_link):
```
Is it possible that the tarball format will be tgz?

Gvaihir replied:
sure, will add that if we add a dependency which is delivered as tgz

commandlinegirl commented on dxcint/dxcint/Logger.py:
```diff
@@ -0,0 +1,45 @@
+    bold_red = '\x1b[31;1m'
+    reset = '\x1b[0m'
+
+    def __init__(self, fmt):
```
Nice formatting of the logs!:)

commandlinegirl commented on dxcint/dxcint/Messenger.py:
```diff
@@ -0,0 +1,60 @@
+            logging.info(f"Analysis {test_name_with_variant} succeeded")
+            self._status.state = Status.FINISHED
+        except dxpy.DXJobFailureError:
+            logging.info(f"Analysis {test_name_with_variant} failed. This failure will verified")
```
will be verified?

Gvaihir replied:
Sorry - ignore the messenger for now, did not implement Messenger to it's completion yet: here's [the error](https://github.com/dnanexus/dxCompiler/blob/3edc6089761c2a54216160eeb67550fdc433820f/dxcint/dxcint/RegisteredTest.py#L183) if I try to use it and forget about incomplete implementation. Messenger class is for asynchronous monitoring of tests, to avoid `wait_on_done()` and long running suites.
But basically it says that the run is failed, and the class that handles these tests has to verify if the failure is expected (like in [`ExpectedFailure`](https://github.com/dnanexus/dxCompiler/blob/3edc6089761c2a54216160eeb67550fdc433820f/dxcint/dxcint/testclasses/ExpectedFailure.py) class). Albeit `ExpectedFailure` is for now implemented non-asynchronously.

commandlinegirl commented on dxcint/dxcint/RegisteredTest.py:
```diff
@@ -0,0 +1,183 @@
+        self._job_id = None
+        self._test_results = None
+        self._executable_type_switch = {
+            "workflow": dxpy.DXWorkflow,
```
We also run GlobalWorkflow (e.g [here](https://github.com/dnanexus/dxCompiler/blob/develop/scripts/run_multiuser_tests.py#L218)), separately in the integration pipeline though https://github.com/dnanexus/dxCompiler/blob/develop/.github/workflows/integration.yml#L71 We don't compile global workflows (and apps) with dxCompiler.

Gvaihir replied:
cool, sure. Will add that no problem - as soon as I start implementing other subclasses of `RegisteredTest`. Compiling or not - will be handled in those subclasses. I'll just override [this method](https://github.com/dnanexus/dxCompiler/blob/3edc6089761c2a54216160eeb67550fdc433820f/dxcint/dxcint/RegisteredTest.py#L99) to skip compilation

commandlinegirl commented on dxcint/dxcint/Terraform.py:
```diff
@@ -0,0 +1,234 @@
+            )
+        return asset_id.get("id")
+
+    def _create_asset_spec(self, language: str) -> Dict:
```
Will this also be used when we build assets during the release?

Gvaihir replied:
I aimed to standardize all dependency delivery and asset building as much as possible, so shouldn't be a problem. Can build a special CLI command for that to only terraform for releases.

commandlinegirl reviewed (changes requested):
Mostly minor stuff

Gvaihir committed:
APPS-1121 Update integration test suite
(feat) cli.py - added docstrings
(fix) README.md - addressing review comments

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) dxfuse.json - updated dependency version
(fix) README.md - addressing review comments

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) README.md - updated docstrings

commandlinegirl commented on dxcint/README.md:
````diff
@@ -15,39 +15,49 @@ self testing.
 ```commandline
 Usage: dxcint integration [OPTIONS] DXC_REPOSITORY_ROOT
 
+  Run integration test or test suire
````
s/suire/suite

Gvaihir replied:
fixed in [554d916](https://github.com/dnanexus/dxCompiler/pull/312/commits/554d916f62d2434d5c1407215255d8993e8997e7)

commandlinegirl reviewed (commented):
This LGTM, with a typo and one minor request.

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) dxfuse.json - update

Gvaihir committed:
APPS-1121 Update integration test suite
(fix) cli.py & README.md - typos

commandlinegirl reviewed (approved):
LGTM, very nice work, thank you for this!

Gvaihir committed:
Merge branch 'develop' into APPS-1121_integration_test_suite
